File: src/lib/mangadex/theme.ts
import { createTheme } from "@svelteuidev/core";

export type Accent = {
	default: string;
	hover: string;
	active: string;
};

export type PrimaryColor = {
	primary: string;
	primary1: string;
	primary2: string;
};

export type StatusColor = {
	red: string;
	green: string;
	yellow: string;
	blue: string;
	grey: string;
	purple: string;
};

export type IndicatorColor = {
	blue: string;
};

export type DangerColor = {
	default: string;
	l1: string;
	l2: string;
};

export type ButtonAccentColor = {
	default: string;
	alternate: string;
};

export type ScrollbarColor = {
	default: string;
	hovered: string;
};

export type Accents = {
	default: Accent;
	l1: Accent;
	l2: Accent;
	l3: Accent;
	l4: Accent;
	l5: Accent;
};

export type Contrast = {
	l1: string;
};

export type MangadexTheme = {
	textColor: string;
	mainBackground: string;
	accents: Accents;
	mid_tone: string;
	contrast: Contrast;
	scrollbar: ScrollbarColor;
	button: ButtonAccentColor;
	primary: PrimaryColor;
	status: StatusColor;
	indication: IndicatorColor;
	danger: DangerColor;
};

export const custom: MangadexTheme = {
	textColor: "#ffffff",
	mainBackground: "#191a1c",
	accents: {
		default: {
			default: "#2c2c2c",
			hover: "#393939",
			active: "#1d1d1d"
		},
		l1: {
			default: "#3d3d3d",
			hover: "#494949",
			active: "#272727"
		},
		l2: {
			default: "#4f4f4f",
			hover: "#686868",
			active: "#222222"
		},
		l3: {
			default: "#585858",
			hover: "#707070",
			active: "#3e3e3e"
		},
		l4: {
			default: "#5d5d5d",
			hover: "#686868",
			active: "#373737"
		},
		l5: {
			default: "#666666",
			hover: "#797979",
			active: "#3b3b3b"
		}
	},
	mid_tone: "#8c8c8c",
	contrast: {
		l1: "#4f4f4f"
	},
	scrollbar: {
		default: "#4f4f4f",
		hovered: "#585858"
	},
	button: {
		default: "#4f4f4f",
		alternate: "#2c2c2c"
	},
	primary: {
		primary: "#ff00c0",
		primary1: "#ff18d8",
		primary2: "#dd00c0"
	},
	status: {
		red: "#ff4040",
		green: "#04d000",
		yellow: "#eeff00",
		blue: "#00c9f5",
		grey: "#9d9d9d",
		purple: "#7d40ff"
	},
	indication: {
		blue: "#4b98f1"
	},
	danger: {
		default: "#ff4040",
		l1: "#db3131",
		l2: "#bf2121"
	}
};

export function buildSvelteUITheme({
	primary,
	status,
	accents,
	scrollbar,
	button,
	indication,
	danger: _danger,
	textColor: color,
	mainBackground: backgroundColor,
	contrast
}: MangadexTheme) {
	const {
		red: statusRed,
		green: statusGreen,
		yellow: statusYellow,
		blue: statusBlue,
		grey: statusGrey,
		purple: statusPurple
	} = status;
	const { default: l0, l1, l2, l3, l4, l5 } = accents;
	const { default: danger, l1: dangerl1, l2: dangerl2 } = _danger;
	return createTheme("mangadex", {
		fonts: {
			standard: "Poppins"
		},
		colors: {
			backgroundColor,
			color,
			primary: primary.primary,
			primary1: primary.primary1,
			primary2: primary.primary2,
			statusRed,
			statusGreen,
			statusYellow,
			statusBlue,
			statusGrey,
			statusPurple,
			accent: l0.default,
			accentHover: l0.hover,
			accentActive: l0.active,
			accentL1: l1.default,
			accentL1Hover: l1.hover,
			accentL1Active: l1.active,
			accentL2: l2.default,
			accentL2Hover: l2.hover,
			accentL2Active: l2.active,
			accentL3: l3.default,
			accentL3Hover: l3.hover,
			accentL3Active: l3.active,
			accentL4: l4.default,
			accentL4Hover: l4.hover,
			accentL4Active: l4.active,
			accentL5: l5.default,
			accentL5Hover: l5.hover,
			accentL5Active: l5.active,
			buttonAccent: button.default,
			buttonAccentAlt: button.alternate,
			scrollbar: scrollbar.default,
			scrollbarHover: scrollbar.hovered,
			indicationBlue: indication.blue,
			constrastL1: contrast.l1,
			danger,
			dangerl1,
			dangerl2
		}
	});
}

File: src-tauri/mangadex/src/store/types/structs/theme.rs
use async_graphql::{InputObject, SimpleObject};
use serde::{Deserialize, Serialize};

pub mod profiles;

#[derive(Debug, Clone, Deserialize, Serialize, SimpleObject, InputObject)]
#[graphql(input_name = "AccentInput")]
pub struct Accent {
    pub default: String,
    pub hover: String,
    pub active: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, SimpleObject, InputObject)]
#[graphql(input_name = "PrimaryColorInput")]
pub struct PrimaryColor {
    pub primary: String,
    pub primary1: String,
    pub primary2: String,
}

impl Default for PrimaryColor {
    fn default() -> Self {
        Self {
            primary: "#ff00c0".into(),
            primary1: "#ff18d8".into(),
            primary2: "#dd00c0".into(),
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, SimpleObject, InputObject)]
#[graphql(input_name = "StatusColorInput")]
pub struct StatusColor {
    pub red: String,
    pub green: String,
    pub yellow: String,
    pub blue: String,
    pub grey: String,
    pub purple: String,
}

impl Default for StatusColor {
    fn default() -> Self {
        Self {
            red: "#ff4040".into(),
            green: "#04d000".into(),
            yellow: "#eeff00".into(),
            blue: "#00c9f5".into(),
            grey: "#9d9d9d".into(),
            purple: "#7d40ff".into(),
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, SimpleObject, InputObject)]
#[graphql(input_name = "IndicatorColorInput")]
pub struct IndicatorColor {
    pub blue: String,
}

impl Default for IndicatorColor {
    fn default() -> Self {
        Self {
            blue: "#4b98f1".into(),
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, SimpleObject, InputObject)]
#[graphql(input_name = "DangerColorInput")]
pub struct DangerColor {
    pub default: String,
    pub l1: String,
    pub l2: String,
}

impl Default for DangerColor {
    fn default() -> Self {
        Self {
            default: "#ff4040".into(),
            l1: "#db3131".into(),
            l2: "#bf2121".into(),
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, SimpleObject, InputObject)]
#[graphql(input_name = "ButtonAccentInput")]
pub struct ButtonAccentColor {
    pub default: String,
    pub alternate: String,
}

impl Default for ButtonAccentColor {
    fn default() -> Self {
        Self {
            default: "#4f4f4f".into(),
            alternate: "#2c2c2c".into(),
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, SimpleObject, InputObject)]
#[graphql(input_name = "ScrollbarColorInput")]
pub struct ScrollbarColor {
    pub default: String,
    pub hovered: String,
}

impl Default for ScrollbarColor {
    fn default() -> Self {
        Self {
            default: "#4f4f4f".into(),
            hovered: "#585858".into(),
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, SimpleObject, InputObject)]
#[graphql(input_name = "AccentsInput")]
pub struct Accents {
    pub default: Accent,
    pub l1: Accent,
    pub l2: Accent,
    pub l3: Accent,
    pub l4: Accent,
    pub l5: Accent,
}

impl Default for Accents {
    fn default() -> Self {
        Self {
            default: Accent {
                default: "#2c2c2c".into(),
                hover: "#393939".into(),
                active: "#1d1d1d".into(),
            },
            l1: Accent {
                default: "#3d3d3d".into(),
                hover: "#494949".into(),
                active: "#272727".into(),
            },
            l2: Accent {
                default: "#4f4f4f".into(),
                hover: "#686868".into(),
                active: "#222222".into(),
            },
            l3: Accent {
                default: "#585858".into(),
                hover: "#707070".into(),
                active: "#3e3e3e".into(),
            },
            l4: Accent {
                default: "#5d5d5d".into(),
                hover: "#686868".into(),
                active: "#373737".into(),
            },
            l5: Accent {
                default: "#666666".into(),
                hover: "#797979".into(),
                active: "#3b3b3b".into(),
            },
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, SimpleObject, InputObject)]
#[graphql(input_name = "ContrastInput")]
pub struct Contrast {
    pub l1: String,
}

impl Default for Contrast {
    fn default() -> Self {
        Self {
            l1: "#4f4f4f".into(),
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, SimpleObject, InputObject)]
#[graphql(input_name = "MangaDexThemeInput")]
pub struct MangaDexTheme {
    pub text_color: String,
    pub main_background: String,
    pub accents: Accents,
    pub mid_tone: String,
    pub contrast: Contrast,
    pub scrollbar: ScrollbarColor,
    pub button: ButtonAccentColor,
    pub primary: PrimaryColor,
    pub status: StatusColor,
    pub indication: IndicatorColor,
    pub danger: DangerColor,
}

impl Default for MangaDexTheme {
    fn default() -> Self {
        Self {
            text_color: "#ffffff".into(),
            main_background: "#191a1c".into(),
            accents: Default::default(),
            mid_tone: "#8c8c8c".into(),
            contrast: Default::default(),
            scrollbar: Default::default(),
            button: Default::default(),
            primary: Default::default(),
            status: Default::default(),
            indication: Default::default(),
            danger: Default::default(),
        }
    }
}
